### ISSUE TITLE
100 Continue is not handled correctly in all cases

### ISSUE THREAD

**Issue by MFranceschi6** [swift-server/async-http-client#461]
Issue related to:
AsyncHttpClient Version 1.6.0 onward

Context:
We interact with a server that always responds first with a 100 continue status code. Before version 1.6 we were able to handle this case with something like this:
```swift
extension EventLoopFuture where Value == HTTPClient.Response {
func handleContinueStatusCode(client: HTTPClient, request: HTTPClient.Request) -> EventLoopFuture<HTTPClient.Response> {

        let resp = self.eventLoop.makePromise(of: HTTPClient.Response.self)
        
        self.whenSuccess { response in
            if response.status == .continue {
                let next = client.execute(request.method, url: "\(request.url)")
                        .handleContinuationStatusCode(client: client, request: request)
                resp.completeWith(next)
            }
            else {
                resp.succeed(response)
            }
        }
...
}
```

After version 1.6 the client started to throw the new `httpEndReceivedAfterHeadWith1xx` error thown here
https://github.com/swift-server/async-http-client/blob/1361eccfb2493a9fdadb2d14675b5e3d6731f3ec/Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift#L531-L538

Now we are exactly in this situation and we changed the code above with:
```swift
extension EventLoopFuture where Value == HTTPClient.Response {
func handleContinueStatusCode(client: HTTPClient, request: HTTPClient.Request) -> EventLoopFuture<HTTPClient.Response> {

        let resp = self.eventLoop.makePromise(of: HTTPClient.Response.self)
        
        self.whenFailure { error in
            if let error =  error as? HTTPClientError, error == .httpEndReceivedAfterHeadWith1xx {
                let next = client.execute(request.method, url: "\(request.url)")
                        .handleContinuationStatusCode(client: client, request: request)
                resp.completeWith(next)
            }
            else {
                resp.fail(error)
            }
        }
...
}
```

But this doesn't work anymore.
As a quirk fix for our case, we found that modifying the handling of the error at line 538 with
`return .failRequest(error, .none)`

The `handleContinueStatusCode` start to work again. But we don't know what other impacts this change has.
For an additional Insight I add here the curl of the request that fails:
```curl
curl --location --request POST 'https://sandbox.cedacri.it/psd2/v1.2/bg/03105/v1/consents' \
--header 'x-request-id: 3c528418-c8dc-4627-95e2-3e05e93411d8' \
--header 'psu-ip-address: 127.0.0.1' \
--header 'tpp-redirect-uri: http://localhost' \
--header 'Content-Type: application/json' \
--data-raw '{"validUntil":"1970-01-01","access":{"availableAccounts":"allAccounts"},"frequencyPerDay":1,"recurringIndicator":false,"combinedServiceIndicator":false}' -v 
```

Swift version:
Swift version 5.5 (swift-5.5-RELEASE)
Target: x86_64-unknown-linux-gnu

**Comment by fabianfett** [comment]
From what I can see this is not actually an issue of AHC, but of SwiftNIO. We should keep the issue open here though.
https://github.com/apple/swift-nio/pull/1330
https://github.com/apple/swift-nio/pull/1422

**Comment by fabianfett** [comment]
This issue hasn't been fully fixed yet.

**Comment by fabianfett** [comment]
@MFranceschi6 We will release a new version of AHC shortly. Once you have updated, you will not need any special handling for your request anymore. It **should** just work, without any retries whatsoever.

Thanks for reporting the issue. If you run into any further issues please reach out!

**Comment by fabianfett** [comment]
Released in 1.6.4

**Comment by MFranceschi6** [comment]
Thank you very much I will let you know ASAP if the release has fixed the issue.
